### ISSUE TITLE
Cannot serialize supplementary characters (Unicode above \uFFFF).

### ISSUE THREAD

**Issue by zawlazaw** [frugalmechanic/fm-serializer#7]
First of all, thanks for this great library!
I stumbled upon an issue when serializing supplementary characters (such as many Emojis).
Background information on dealing with this kind of characters on the JVM can be found here:
 http://www.oracle.com/us/technologies/java/supplementary-142654.html

The following code breaks, and I am unable to fix this or to find a proper workaround:

```
import fm.serializer.protobuf.Protobuf

object Example extends App {
  val s = "\uD83D\uDCA5" // same as "💥"
  val bytes = Protobuf.toBytes(s)
  val t = Protobuf.fromBytes[String](bytes)
  println(s + " != " + t) // gives 💥 != ��
  assert(s == t) // fails
}

```

**Comment by tpunder** [comment]
Good catch!  This is clearly a lack of understanding on my part about supplementary characters in Java.  The linked article does a good job of explaining it though.  Thanks for that.

Looks like support needs to be added in https://github.com/frugalmechanic/fm-serializer/blob/master/src/main/scala/fm/serializer/FMByteArrayOutputStream.scala that checks for these supplementary characters and writes them out in proper UTF-8 encoding.

**Comment by zawlazaw** [comment]
This seems to be equivalent to supporting 4-byte-UTF8-characters, which is currently explicitly excluded in the file you linked (`MAX_UTF8_CHAR_BYTES = 3`).
Would the reading side require changes, too?
Currently, as a workaround, I simply drop any such "special chars" from the string, which is easiest to achieve via `val cleaned = s.replaceAll("[^\u0000-\uffff]", "")`

**Comment by tpunder** [comment]
Here is an initial pass at support:  https://github.com/frugalmechanic/fm-serializer/commit/a9c17e5a402db9fee3ea5a68846a9c7c9227a604

It's published as `0.11.0-SNAPSHOT`

**Comment by zawlazaw** [comment]
Wow, that's fast... All code changes seem to be plausible (although I did not double-check all details). I also tried to break it after the fix, but I was not successful - which is good :) . So it seems to be fine!


    import fm.serializer.protobuf.Protobuf
  
    object Example extends App {
  
      val examples = Seq(
        "",
        "\u0000",
        "💥",
        "\uD83D\uDCA5",
        "\uD83D\uDCA5"*99,
        "\u0000💥\u0000"*99,
        "ᒪ𝜎ⲅ𝚎ｍ 𝒊ϱ𝔰𝒖ｍ 𝔡ﮬl𝜊𝘳 𝘴і𝙩 𝕒ｍ𝖊𝚝, 𝘤ﮨℼ𐑈е𝞃𝓮𝕥𝛖𝘳 𝒔а𝗱ꙇ𝒑𝗌𝒸і𝑛ᶃ еlι𝘵ⲅ, 𝖘𝒆Ꮷ 𝕕ı⍺ｍ ℼ۵𝝕𝑢ｍɣ 𝖊ℹ𝓻ｍ𐐬𝚍 𝖙ｅｍ𝓹ﮨ𝕣 𝕚ռ𝕧𝜾𝖉ｕп𝐭 𝔲т l𝓪𝕓օᴦ𝚎 е𝑡 𝓭ﮩlﮦⲅ𝓮 ｍɑ𝐠𝞏𝒂 𝖆l𝑖գ𝙪𝕪𝐚ｍ 𝑒𝔯ɑ𝛕, 𝓼𝖊𝙙 ꓒ𝗶𝐚ｍ 𝖛𝞸l𝚞𝛠τ𝙪𝝰. 𝗔𝛕 ⋁𝐞𝙧𝑜 𝚎ⲟ𝗌 𝐞ᴛ 𝜶𝖈ᴄ𝘶ѕ𝚊ｍ 𝘦𝗍 𝗃ｕꜱт𝗼 𝔡ц𝔬 ꓒ𝔬l𝞼𝓻ⅇ𝙨 𝗲𝔱 е𝒂 𝓻𝓮𝔟𝕦ｍ. 𝕾𝙩𝓮𝘁 𝙘l𝓲𝘵𝓪 𝝹𝛂𝘀ⅾ 𝖌𝝊Ƅ℮ᴦ𝙜г𝒆𝕟, 𝜛𝕠 𝕤𝖊𝐚 𝕥𝙖𝖐ｉｍ𝙖𝘁𝐚 𝐬𝙖𝚗𝓬𝗍𝙪ѕ 𝐞𝘀𝜏 𐑃ᴑᴦ𝔢ｍ ͺ𝛠𝐬𝐮ｍ 𝘥၀lﮩ𝓻 ꜱ𝑖𝒕 𝖺ｍ𝕖τ. 𝕷𝞸𝓻𝖊ｍ 𝚤𝓅𝑠𝓾ｍ 𝐝੦l૦ᴦ 𝖘ͺ𝖙 𝔞ｍ𝑒𝞃, 𝒄ہｎ𝑠℮𝓽𝘦𝙩𝑢𝑟 𝑠𝗮𝚍𝚤𝝆𝒔𝓬і𝚗𝐠 ｅl𝚤𝑡𝖗, ꜱ𝑒𝔡 Ꮷ𝗶𝘢ｍ 𝚗ﮬ𝛑𝗎ｍ𝘺 𝑒ι𝘳ｍﻪᑯ т𝔢ｍ𝝔𝓸г ɩℼ⋁ℹ𝗱𝖚𝜋𝔱 𝜐𝘁 l𝞪𝘣၀𝕣𝐞 𝓮𝘁 ｄ𝞼lﮬ𝓇𝖾 ｍ𝜶𝔤𝞏𝝰 ⍺lⅈԛ𝙪ү𝑎ｍ 𝗲гɑ𝞃, 𝘀𝙚ᑯ 𝒅⍳𝘢ｍ ט𝒐l𝒖𝛒𝗍𝗎𝜶. 𝔄𝖙 𝗏е𝚛𝝄 ℯﮫƽ 𝖾ｔ 𝝰𝗰ϲ𝔲ѕａｍ ℯｔ ϳ𝝊𐑈𝗍𝛐 ｄ𝗎ه 𝖉οlσг𝙚𝙨 𝚎𝘵 ℮𝞪 𝑟𝚎𝖇𝑢ｍ. 𝕾𝒕𝙚𝖙 𝗰l𝔦𝓉𝖺 𝜘𝞪ƽ𝕕 𝗴𝜐Ꮟℯｒƍ𝐫еℼ, 𝕟σ ѕｅ𝛂 𝕥𝒂𝔨𝚤ｍ𝒶𝐭𝒶 𝘴⍺𝔫ϲｔ𝗎𝑠 𝗲𝚜𝞽 ⳑه𝔯𝔢ｍ ⍳𝒑ꜱ𝒖ｍ 𝓭ﮨlھ𝗿 𝒔ꙇ𝚝 𝚊ｍ𝗲τ.",
        "ᒪ\uD835\uDF0Eⲅ\uD835\uDE8Eｍ \uD835\uDC8Aϱ\uD835\uDD30\uD835\uDC96ｍ \uD835\uDD21ﮬl\uD835\uDF0A\uD835\uDE33 \uD835\uDE34і\uD835\uDE69 \uD835\uDD52ｍ\uD835\uDD8A\uD835\uDE9D, \uD835\uDE24ﮨℼ\uD801\uDC48е\uD835\uDF83\uD835\uDCEE\uD835\uDD65\uD835\uDED6\uD835\uDE33 \uD835\uDC94а\uD835\uDDF1ꙇ\uD835\uDC91\uD835\uDDCC\uD835\uDCB8і\uD835\uDC5Bᶃ еlι\uD835\uDE35ⲅ, \uD835\uDD98\uD835\uDC86Ꮷ \uD835\uDD55ı⍺ｍ ℼ۵\uD835\uDF55\uD835\uDC62ｍɣ \uD835\uDD8Aℹ\uD835\uDCFBｍ\uD801\uDC2C\uD835\uDE8D \uD835\uDD99ｅｍ\uD835\uDCF9ﮨ\uD835\uDD63 \uD835\uDD5Aռ\uD835\uDD67\uD835\uDF3E\uD835\uDD89ｕп\uD835\uDC2D \uD835\uDD32т l\uD835\uDCEA\uD835\uDD53օᴦ\uD835\uDE8E е\uD835\uDC61 \uD835\uDCEDﮩlﮦⲅ\uD835\uDCEE ｍɑ\uD835\uDC20\uD835\uDF8F\uD835\uDC82 \uD835\uDD86l\uD835\uDC56գ\uD835\uDE6A\uD835\uDD6A\uD835\uDC1Aｍ \uD835\uDC52\uD835\uDD2Fɑ\uD835\uDED5, \uD835\uDCFC\uD835\uDD8A\uD835\uDE59 ꓒ\uD835\uDDF6\uD835\uDC1Aｍ \uD835\uDD9B\uD835\uDFB8l\uD835\uDE9E\uD835\uDEE0τ\uD835\uDE6A\uD835\uDF70. \uD835\uDDD4\uD835\uDED5 ⋁\uD835\uDC1E\uD835\uDE67\uD835\uDC5C \uD835\uDE8Eⲟ\uD835\uDDCC \uD835\uDC1Eᴛ \uD835\uDF36\uD835\uDD88ᴄ\uD835\uDE36ѕ\uD835\uDE8Aｍ \uD835\uDE26\uD835\uDDCD \uD835\uDDC3ｕꜱт\uD835\uDDFC \uD835\uDD21ц\uD835\uDD2C ꓒ\uD835\uDD2Cl\uD835\uDFBC\uD835\uDCFBⅇ\uD835\uDE68 \uD835\uDDF2\uD835\uDD31 е\uD835\uDC82 \uD835\uDCFB\uD835\uDCEE\uD835\uDD1F\uD835\uDD66ｍ. \uD835\uDD7E\uD835\uDE69\uD835\uDCEE\uD835\uDE01 \uD835\uDE58l\uD835\uDCF2\uD835\uDE35\uD835\uDCEA \uD835\uDF79\uD835\uDEC2\uD835\uDE00ⅾ \uD835\uDD8C\uD835\uDF4AƄ℮ᴦ\uD835\uDE5Cг\uD835\uDC86\uD835\uDD5F, \uD835\uDF1B\uD835\uDD60 \uD835\uDD64\uD835\uDD8A\uD835\uDC1A \uD835\uDD65\uD835\uDE56\uD835\uDD90ｉｍ\uD835\uDE56\uD835\uDE01\uD835\uDC1A \uD835\uDC2C\uD835\uDE56\uD835\uDE97\uD835\uDCEC\uD835\uDDCD\uD835\uDE6Aѕ \uD835\uDC1E\uD835\uDE00\uD835\uDF0F \uD801\uDC43ᴑᴦ\uD835\uDD22ｍ ͺ\uD835\uDEE0\uD835\uDC2C\uD835\uDC2Eｍ \uD835\uDE25၀lﮩ\uD835\uDCFB ꜱ\uD835\uDC56\uD835\uDC95 \uD835\uDDBAｍ\uD835\uDD56τ. \uD835\uDD77\uD835\uDFB8\uD835\uDCFB\uD835\uDD8Aｍ \uD835\uDEA4\uD835\uDCC5\uD835\uDC60\uD835\uDCFEｍ \uD835\uDC1D੦l૦ᴦ \uD835\uDD98ͺ\uD835\uDD99 \uD835\uDD1Eｍ\uD835\uDC52\uD835\uDF83, \uD835\uDC84ہｎ\uD835\uDC60℮\uD835\uDCFD\uD835\uDE26\uD835\uDE69\uD835\uDC62\uD835\uDC5F \uD835\uDC60\uD835\uDDEE\uD835\uDE8D\uD835\uDEA4\uD835\uDF46\uD835\uDC94\uD835\uDCECі\uD835\uDE97\uD835\uDC20 ｅl\uD835\uDEA4\uD835\uDC61\uD835\uDD97, ꜱ\uD835\uDC52\uD835\uDD21 Ꮷ\uD835\uDDF6\uD835\uDE22ｍ \uD835\uDE97ﮬ\uD835\uDED1\uD835\uDDCEｍ\uD835\uDE3A \uD835\uDC52ι\uD835\uDE33ｍﻪᑯ т\uD835\uDD22ｍ\uD835\uDF54\uD835\uDCF8г ɩℼ⋁ℹ\uD835\uDDF1\uD835\uDD9A\uD835\uDF0B\uD835\uDD31 \uD835\uDF10\uD835\uDE01 l\uD835\uDFAA\uD835\uDE23၀\uD835\uDD63\uD835\uDC1E \uD835\uDCEE\uD835\uDE01 ｄ\uD835\uDFBClﮬ\uD835\uDCC7\uD835\uDDBE ｍ\uD835\uDF36\uD835\uDD24\uD835\uDF8F\uD835\uDF70 ⍺lⅈԛ\uD835\uDE6Aү\uD835\uDC4Eｍ \uD835\uDDF2гɑ\uD835\uDF83, \uD835\uDE00\uD835\uDE5Aᑯ \uD835\uDC85⍳\uD835\uDE22ｍ ט\uD835\uDC90l\uD835\uDC96\uD835\uDED2\uD835\uDDCD\uD835\uDDCE\uD835\uDF36. \uD835\uDD04\uD835\uDD99 \uD835\uDDCFе\uD835\uDE9B\uD835\uDF44 ℯﮫƽ \uD835\uDDBEｔ \uD835\uDF70\uD835\uDDF0ϲ\uD835\uDD32ѕａｍ ℯｔ ϳ\uD835\uDF4A\uD801\uDC48\uD835\uDDCD\uD835\uDED0 ｄ\uD835\uDDCEه \uD835\uDD89οlσг\uD835\uDE5A\uD835\uDE68 \uD835\uDE8E\uD835\uDE35 ℮\uD835\uDFAA \uD835\uDC5F\uD835\uDE8E\uD835\uDD87\uD835\uDC62ｍ. \uD835\uDD7E\uD835\uDC95\uD835\uDE5A\uD835\uDD99 \uD835\uDDF0l\uD835\uDD26\uD835\uDCC9\uD835\uDDBA \uD835\uDF18\uD835\uDFAAƽ\uD835\uDD55 \uD835\uDDF4\uD835\uDF10Ꮟℯｒƍ\uD835\uDC2Bеℼ, \uD835\uDD5Fσ ѕｅ\uD835\uDEC2 \uD835\uDD65\uD835\uDC82\uD835\uDD28\uD835\uDEA4ｍ\uD835\uDCB6\uD835\uDC2D\uD835\uDCB6 \uD835\uDE34⍺\uD835\uDD2Bϲｔ\uD835\uDDCE\uD835\uDC60 \uD835\uDDF2\uD835\uDE9C\uD835\uDFBD ⳑه\uD835\uDD2F\uD835\uDD22ｍ ⍳\uD835\uDC91ꜱ\uD835\uDC96ｍ \uD835\uDCEDﮨlھ\uD835\uDDFF \uD835\uDC94ꙇ\uD835\uDE9D \uD835\uDE8Aｍ\uD835\uDDF2τ."
      )
  
      for (s <- examples) {
  
        println(s)
        println(s.length)
  
        assert(s == Protobuf.fromBytes[String](Protobuf.toBytes(s)))
        assert(s == Protobuf.fromBase64String[String](Protobuf.toBase64String(s)))
        assert(s == Protobuf.fromBase64URLSafeString[String](Protobuf.toBase64URLSafeString(s)))
  
        assert(s == JSON.fromBytes[String](JSON.toBytes(s)))
        assert(s == JSON.fromJSON[String](JSON.toMinimalJSON(s)))
        assert(s == JSON.fromJSON[String](JSON.toPrettyJSON(s)))
        assert(s == JSON.fromJSON[String](JSON.toJSON(s)))
  
        assert(s == BSON.fromBsonBytes[String](BSON.toBsonBytes(s)))
        assert(s == BSON.fromBsonDocument[String](BSON.toBsonDocument(s)))
        assert(s == BSON.fromBsonDocument[String](BSON.toRawBsonDocument(s)))
      }
    }
btw: I checkout out master, since `0.11.0-SNAPSHOT` does not seem to show up on Maven central, or am I missing something?

**Comment by tpunder** [comment]
I've added your additional examples as tests  :)

`0.11.0-SNAPSHOT` is published to Sonatype Snapshots (https://oss.sonatype.org/content/repositories/snapshots/ or `Resolver.sonatypeRepo("snapshots")` in SBT).

**Comment by tpunder** [comment]
This has been published to Maven Central as version `0.11.0`
